File: src/DesignPattern/Behavioral/Observer/Observer-go/observer.go
package main

import "fmt"

type observer interface {
	Update()
}

// ConcreteObserver ...
type ConcreteObserver struct {
	subject       *Subuject
	observerState int
}

// Update ...
func (p *ConcreteObserver) Update() {
	if p.observerState != p.subject.GetState() {
		fmt.Println("变化了")
	}
	p.observerState = p.subject.GetState()
}

// Subuject ...
type Subuject struct {
	observers []observer
	data      int
}

// Attact ...
func (p *Subuject) Attact(obs observer) {
	p.observers = append(p.observers, obs)
}

// SetState ...
func (p *Subuject) SetState(state int) {
	p.data = state
	p.Notify()
}

// GetState ...
func (p *Subuject) GetState() int {
	return p.data
}

// Notify ...
func (p *Subuject) Notify() {
	for _, v := range p.observers {
		v.Update()
	}

}

File: src/DesignPattern/Behavioral/Observer/observer.ts
export interface Observer {
    Update()
};

export class ConcreteObserver implements Observer {
    constructor(subj: Subject) {
        this.subject = subj;
    }
    private observerState: number;
    private subject: Subject;
    private quite(oldData: number, newData: number) {
        if (oldData > newData) {
            console.log("变小了");
        }
        else if (oldData < newData) {
            console.log("变大了");
        } else {
            console.log("没变化");
        }
    };
    public Update() {
        let newData: number = this.subject.GetState();
        this.quite(this.observerState, newData);
        this.observerState = newData;
    }
};


export class Subject {
    constructor() {
        this.observers = new Array<Observer>();
    }
    protected observers: Observer[];
    protected state: number;
    public Attach(observer: Observer) {
        this.observers.push(observer);
    }
    // 通知观察者
    public Notufy() {
        for (const iterator of this.observers) {
            iterator.Update()
        }
    }
    public GetState(): number {
        return this.state;
    }
    public SetState(data: number) {
        this.state = data;
        this.Notufy();
    }
};

function main() {
    const subject = new Subject();
    
    // 创建观察者
    const observer = new ConcreteObserver(subject);
    
    // 添加观察者
    subject.Attach(observer);

    for (let i: number = 0; i < 5; i++) {
        subject.SetState(i);
    }
    for (let i: number = 5; i > 0; i--) {
        subject.SetState(i);
    }
}